[Intel GPU] oneDNN GPU GEMM support

ZhiweiYan-96 opened the pull request:
# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `addmm`, `baddmm` is defined in `Blas.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info processing in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise` in #117824. 

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118064
* #117824
* #117529
* #117512
* __->__ #117202
* #117112
* #117098



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

ZhiweiYan-96 committed:
[Intel GPU] oneDNN GPU GEMM support

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

This PR contains oneDNN GEMM support, which consists of oneDNN glue code `Matmul.h` defined `aten/src/ATen/native/mkldnn/xpu/detail` directory. Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

This PR contains oneDNN GEMM support, which consists of oneDNN glue code `Matmul.h` defined `aten/src/ATen/native/mkldnn/xpu/detail` directory. Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

This PR contains oneDNN GEMM support, which consists of oneDNN glue code `Matmul.h` defined `aten/src/ATen/native/mkldnn/xpu/detail` directory. Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

This PR contains oneDNN GEMM support, which consists of oneDNN glue code `Matmul.h` defined `aten/src/ATen/native/mkldnn/xpu/detail` directory. Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

EikanWang commented:
@ZhiweiYan-96 , please fix the linter issue.

gujinghui commented on aten/src/ATen/native/mkldnn/xpu/BlasImpl.h:
```diff
@@ -0,0 +1,518 @@
+#include <ATen/ATen.h>
```
should we merge this file into Blas.cpp to align with pytorch?

ZhiweiYan-96 replied:
`BlasImpl.h` would be used more than `Blas.cpp`. It would also be used for in `xpu/Linear.cpp` for linear pointwise fusion, it's more like a common utils. I recommend we may keep it :)

EikanWang replied:
Please update your PR description to describe the relationship between `BlasImpl.h` and `Blas.cpp`

jgong5 replied:
What functions are being used by `Blas.cpp`? I might miss something but couldn't find the usage there.

ZhiweiYan-96 replied:
@jgong5  Thanks for the reminding.  The BlasImpl.h is not used in Blas.cpp. I'd move it into Linear fusion pr

xiaolil1 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

xiaolil1 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[draft][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[draft][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[draft][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[draft][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[draft][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[draft][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[draft][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[draft][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

EikanWang commented on aten/src/ATen/native/mkldnn/xpu/Blas.cpp:
```diff
@@ -0,0 +1,422 @@
+#include <ATen/native/mkldnn/xpu/BlasImpl.h>
+#include <torch/library.h>
+
+namespace at {
+namespace native {
+namespace xpu {
```
Since the code is under `mkldnn`, may I know why the namespace does not contain `onednn`?

ZhiweiYan-96 replied:
I choose this is because `at::native::xpu` is aligned with `at::native::cuda` and `at::native::cpu` namespace, which is classified through device backend.
`at::native::onednn` would be a specific namespace for onednn related operators, which is aligned to `at::native::cudnn`

EikanWang commented on caffe2/CMakeLists.txt:
```diff
@@ -676,6 +680,8 @@ if(USE_CUDA)
   set_source_files_properties(${TORCH_SRC_DIR}/csrc/jit/codegen/cuda/interface.cpp PROPERTIES COMPILE_FLAGS "-DUSE_CUDA=1")
 endif()
 
+
+
```
```suggestion
```

ZhiweiYan-96 replied:
has removed

EikanWang commented on test/test_xpu.py:
```diff
@@ -127,5 +137,173 @@ def test_generator(self):
+        }
+    )
+    @dtypes(*floating_and_complex_types_and(torch.half))
+    def test_addmm(self, device, dtype):
```
Please reuse CPU and CUDA test case.

leizhenyuan committed:
Update on "[WIP][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info process in matmul related operators.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info process in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info process in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info process in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info process in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info process in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info process in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[WIP][Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info process in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

EikanWang commented:
@albanD , we are facing the pr-sanity-check failure due to ~1170 line of code for test cases. I'm wondering if you can help review this PR and decide if we can ignore the pr-sanity-check.

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info process in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 commented on test/xpu/test_gemm.py:
hi @jgong5 , the created the code is directly modified from cuda&cpu cases in `test_linalg.py`, where we change the dtype support and remove some cuda specific code. This might not be optimal, but it is fast for us checking the issue in code. Also, @EikanWang is proposing a better way to resuing existing cuda&cpu cases.

jgong5 replied:
Do you mean these tests are temporary and will be removed later? I'm fine if that is so.

ZhiweiYan-96 replied:
Yes, I will use the technique Eikan propose. UT in current will be removed then.

EikanWang replied:
> There are existing UTs for CPU and CUDA. Can we reuse them instead of creating our own?

So far, it is difficult to reuse these cases directly due to the XPU's limited aten operations support and torch test infrastructure. We are trying to figure out an elegant solution, as I describe below.

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `addmm`, `baddmm` is defined in `Blas.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info processing in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise` in #117824. 

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 commented:
The `BlasImpl.h` has been removed in this PR due to no caller in `Blas.cpp`, I would merge `BlasImpl.h` into `Linear.cpp` in #117824

albanD reviewed (commented):
I'm not sure it's ok to skip? Most of the line count seems to come from the manual testing of the op. Why is this not done through OpInfo as we do for all other backends?

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `addmm`, `baddmm` is defined in `Blas.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info processing in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise` in #117824. 

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `addmm`, `baddmm` is defined in `Blas.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info processing in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise` in #117824. 

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

EikanWang commented:
> I'm not sure it's ok to skip? Most of the line count seems to come from the manual testing of the op. Why is this not done through OpInfo as we do for all other backends?

Hi @albanD , we will utilize `OpInfo` as well. Beyond the `OpInfo`, we found that PyTorch tests also provide the dedicated cases for gemm - `test_linalg.py` and some cases are not covered by `OpInfo`. And the `OpInfo` change will impact all the devices but not XPU only. So, we first ported these cases from `test_linalg.py`.

albanD commented:
But the tests in test_linalg.py use the device-generic framework. If you want to re-use them, please add XPU as a device there so that these tests run with XPU.

EikanWang commented:
> But the tests in test_linalg.py use the device-generic framework. If you want to re-use them, please add XPU as a device there so that these tests run with XPU.

XPU only supports limited aten operations so far. To avoid not applicable case failures, we only enable XPU device test cases by setting `only_for` to `xpu` and port applicate test cases like https://github.com/pytorch/pytorch/blob/main/test/test_xpu.py#L231. 

There are two options in my mind.

- Add `@skipIfXPU` for cases that are not applicable. The change may be intrusive as we only support limited operations
- Add another test case file for `linalg` for XPU and inherit this test case `TestLinalg` class and only run the applicable test case.

@albanD

EikanWang commented:
> But the tests in test_linalg.py use the device-generic framework. If you want to re-use them, please add XPU as a device there so that these tests run with XPU.

@albanD , I just submitted a PR to elaborate on the idea https://github.com/pytorch/pytorch/pull/122953. FYI

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `addmm`, `baddmm` is defined in `Blas.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info processing in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise` in #117824. 

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `addmm`, `baddmm` is defined in `Blas.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info processing in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise` in #117824. 

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

ZhiweiYan-96 committed:
Update on "[Intel GPU] oneDNN GPU GEMM support"

# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

Aten operators like `addmm`, `baddmm` is defined in `Blas.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.  `BlasImpl.h` is a header file that provide helper for handling shape info processing in matmul related operators. It would not only help basic GEMM operator like `addmm, baddmm` but also help fusion operators used in `torch.compile` like `linear_pointwise` in #117824. 

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

malfet commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge
